### PR TITLE
[dv/tool] Add support to choose sub-cfgs in dvsim.py

### PR DESCRIPTION
**Pull request by cindychip**
Add a parse_arg for user to choose which dut to run in a master cfg that
contains more than one sub-cfgs.

Signed-off-by: Cindy Chen <chencindy@google.com>

### CODE REVIEW

File: util/dvsim/FlowCfg.py [hunk 1]
```diff
@@ -89,7 +90,8 @@ def __post_init__(self):
         # Run some post init checks
         if not self.is_master_cfg:
             # Check if self.cfgs is a list of exactly 1 item (self)
-            if not (len(self.cfgs) == 1 and self.cfgs[0].name == self.name):
+            # Or has not cfg because the command line set to run certain duts
```
Comment (msfschaffner) [inline util/dvsim/FlowCfg.py]: maybe change the comment here to something like this:
```python
# This is not a master config, hence we can only have a maximum of one cfg item associated with it (self). 
# We may have zero items here, if only a subset of cfgs has been selected to be run via the command line switch -dut
```

File: util/dvsim/dvsim.py [hunk 2]
```diff
@@ -131,6 +131,13 @@ def main():
                         metavar="vcs|xcelium|ascentlint|...",
                         help="Override the tool that is set in hjson file")
 
+    parser.add_argument("-duts",
+                        nargs="*",
+                        default=[],
+                        metavar="cfg1_dut, cfg2_dut, cfg3_dut, ...",
+                        help="""Indicate which dut to run in a master cfg.
+                In default without this arg, the tool will run all the duts in the cfg.""")
```
Comment (msfschaffner) [inline util/dvsim/dvsim.py]: Maybe slightly rephrase this as:
```
Specifies which dut cfg(s) of the master cfg shall be processed. 
If this switch is not specified, dvsim will process all the dut cfgs specified in the master cfg.
```

File: util/dvsim/FlowCfg.py [hunk 3]
```diff
@@ -36,6 +36,7 @@ def __init__(self, flow_cfg_file, proj_root, args):
         self.scratch_root = args.scratch_root
         self.branch = args.branch
         self.job_prefix = args.job_prefix
+        self.duts = args.duts
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:39]: Change this to `select_cfgs`.
Also, create a local copy in FlowCfg instance rather than pass the list handle from `args`.   
`select_cfgs` should be treated very similar to `items`.

File: util/dvsim/FlowCfg.py [hunk 4]
```diff
@@ -243,11 +245,18 @@ def resolve_hjson_raw(self, hjson_dict):
                     cfg_file = subst_wildcards(entry,
                                                self.__dict__,
                                                ignore_error=True)
+                    # skip sub-cfgs that are not in the list
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:248]: Prefer to not change the `resolve_hjson_raw` function since it is already quite complicated. 

Instead, add `prune_selected_cfgs()` and call it at the start of `create_deploy_objects()`. 

`prune_selected_cfgs()` will return immediately if not a master cfg list or if `--select-cfgs` is None. If not, then it will go through the list of `self.cfgs` and remove the ones not selected via `--select-cfgs` switch. This IMO is a lot easier to implement.  

This way, in all other functions that loop over `self.cfgs`, it will only loop over the cfgs selected on the cmd line.

File: util/dvsim/FlowCfg.py [hunk 5]
```diff
@@ -89,7 +90,8 @@ def __post_init__(self):
         # Run some post init checks
         if not self.is_master_cfg:
             # Check if self.cfgs is a list of exactly 1 item (self)
-            if not (len(self.cfgs) == 1 and self.cfgs[0].name == self.name):
+            # Or has not cfg because the command line set to run certain duts
+            if not ((len(self.cfgs) == 1 and self.cfgs[0].name == self.name) or len(self.cfgs) == 0):
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:94]: This change should not be required if my suggestion below is implemented.

File: util/dvsim/FlowCfg.py (hunk 4 above)
Comment (cindychip) [reply util/dvsim/FlowCfg.py:248]: I see, thanks Sri for the detailed suggestion. I was removing them in `parse_flow_cfg` because in `SimCfg.py` function `__init__`, we are checking the modes for all self.cfgs and printing out the scratch directories.
But I think your idea is cleaner. I will try to clean up the `__init__` function. Thanks!

File: util/dvsim/FlowCfg.py [hunk 6]
```diff
@@ -392,6 +394,19 @@ def print_list(self):
...
 
+    # function to prune only selected cfgs to build and run
+    # it will return if the object is not a master_cfg or -select_cfgs is empty
+    def prune_selected_cfgs(self):
+        if self.is_master_cfg == 0 or self.select_cfgs == []:
+            return
+        else:
+            remove_cfgs = []
+            for item in self.cfgs:
+                if item.dut not in self.select_cfgs:
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:405]: I think this should be `item.name` instead.

File: util/dvsim/FlowCfg.py [hunk 7]
```diff
@@ -392,6 +394,19 @@ def print_list(self):
         for item in self.cfgs:
             item._print_list()
 
+    # function to prune only selected cfgs to build and run
+    # it will return if the object is not a master_cfg or -select_cfgs is empty
+    def prune_selected_cfgs(self):
+        if self.is_master_cfg == 0 or self.select_cfgs == []:
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:400]: Perhaps `not` can be used here instead of `== 0 / ==[]`.

File: util/dvsim/dvsim.py [hunk 8]
```diff
@@ -132,6 +132,14 @@ def main():
                         metavar="vcs|xcelium|ascentlint|dc|...",
                         help="Override the tool that is set in hjson file")
 
+    parser.add_argument("-select_cfgs",
+                        nargs="*",
+                        default=[],
+                        metavar="cfg1_dut, cfg2_dut, cfg3_dut, ...",
+                        help="""Specifies which cfg(s) of the master cfg shall be processed.
+                If this switch is not specified, dvsim will process all the dut cfgs specified in
+                the master cfg.""")
```
Comment (sriyerg) [inline util/dvsim/dvsim.py:141]:
```suggestion
                        metavar="cfg1, cfg2, cfg3, ...",
                        help="""Specifies which cfg(s) of the master cfg shall be processed.
                If this switch is not specified, dvsim will process all the cfgs specified in
                the master cfg list.""")
```

File: util/dvsim/FlowCfg.py (hunk 7 above)
Comment (cindychip) [reply util/dvsim/FlowCfg.py:400]: Thanks Sri, yup, `not` works here. Updated as you recommend.

File: util/dvsim/FlowCfg.py (hunk 6 above)
Comment (cindychip) [reply util/dvsim/FlowCfg.py:405]: I see, thanks Sri. I updated to `name`.

File: util/dvsim/FlowCfg.py [hunk 9]
```diff
@@ -402,9 +417,12 @@ def _create_deploy_objects(self):
     def create_deploy_objects(self):
         '''Public facing API for _create_deploy_objects().
         '''
+        self.prune_selected_cfgs()
         if self.is_master_cfg:
             self.deploy = []
             for item in self.cfgs:
+                log.info("[scratch_dir]: [%s]: [%s]", item.name,
```
Comment (sriyerg) [inline util/dvsim/FlowCfg.py:424]: Looks like this will no longer be printed if you pass a non-master cfg. See if there is a better place to put it.

File: util/dvsim/SimCfg.py [hunk 10]
```diff
@@ -126,9 +126,6 @@ def __init__(self, flow_cfg_file, proj_root, args):
 
         # Stuff below only pertains to individual cfg (not master cfg).
         if not self.is_master_cfg:
-            # Print info
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:129]: One possible option is to guard this with:
```python 
      if not self.select_cfgs or self.name in self.select_cfgs:
```

Comment (cindychip) [reply util/dvsim/SimCfg.py:129]: Thanks Sri for the suggestion. I used your condition and applied it to all the functions behind the `log.info`. Please take a look.

Comment (sriyerg) [reply util/dvsim/SimCfg.py:129]: This is definitely quite confusing. Please add TODO to fix this in a better way later.

Comment (cindychip) [reply util/dvsim/SimCfg.py:129]: Sounds good, will do.
